TA-Grant1101/g2g
# Refactor of backend to add JWT

*Important - you will need to drop your customer table. You will now create two new tables: Customer and Authorities*


Added:
JWT functionality
updated project config
Updated dependencies in gradle

Removed: 
Basic Spring Security 

How to test:

1. Drop your Customer table and start the program
2. add the following SQL to your database:
 INSERT INTO customer (name,email, pwd, role,create_dt)
 VALUES ('Happy','happy@example.com', '$2y$12$oRRbkNfwuR8ug4MlzH5FOeui.//1mkd.RsOAJMbykTSupVy.x/vb2', 'admin',CURDATE());

INSERT INTO authorities (customer_id, name)   VALUES (1, 'ROLE_USER')
INSERT INTO authorities (customer_id, name)   VALUES (1, 'ROLE_ADMIN')

*now the customer table should have one entry for our customer Happy, and authorities table should have two rows- both linking to Happy. 

3. Navigate to Postman and test that the permitted routes work. (http://localhost:8080/hello, http://localhost:8080/api/books/home, http://localhost:8080/api/books/{id} whatever you have ex: 1, 4, 5 etc. If you have book objects in your database, you should be able to see the object in the response body. 
4. Next, lets see if the JWT is generating. do a GET request to http://localhost:8080/user. 
-Select authorization->Basic authorization. Enter the credentials for the user you have made: happy@example.com and pwd: 12345.
![image](https://github.com/TA-Grant1101/g2g/assets/132425270/89901579-fae0-4433-9b2a-148131bd80b7)
-In the response, select **headers** and locate the header "Authorization". there should be a long string of characters (that's our token!)
![image](https://github.com/TA-Grant1101/g2g/assets/132425270/7eb5e53a-2aef-4d4f-b943-7e9bb357d772)
-go back up to the request section and navigate to headings in there->add a new heading called Authorization and paste the token as the value 
![image](https://github.com/TA-Grant1101/g2g/assets/132425270/dc7afff3-d3f1-40e0-bb15-c54366ab719a)
-next, go back to the Authorization tab and **remove** the basic authorization. Select NO AUTH ( the token should now take care of this part)
![image](https://github.com/TA-Grant1101/g2g/assets/132425270/58c8577c-7e4a-40c1-9ff5-435db00b83a6)
-now navigate to a route that is locked behind security: (http://localhost:8080/testing. You should be able to see the string in the body:
![image](https://github.com/TA-Grant1101/g2g/assets/132425270/cf92d1d1-6fcb-4f0e-9d1d-aa982d3ad5e8)

WOOO now for the POST tests!
-navigate to http://localhost:8080/api/books/add. Paste a JSON object of a book to add to the database: (here's a sample if you need it)
{
    "bookKey": "/works/OL23204W",
    "title": "Of Mice and Men",
    "bookCover": 14319003,
    "author": "John Steinbeck",
    "firstPublishYear": 1937,
    "averageRating": 3.986607,
    "numberOfReviews": 224,
    "subject": [
        "contemporary fiction",
        "literary fiction",
        "classics",
        "Labor Trilogy",
        "psychological fiction",
        "western stories",
        "Fiction",
        "Classic Literature",
        "Fiction, general",
        "West (u.s.), fiction"
      ],
    "isAvailable": true
}
we're hoping to see that beautiful 200 response:
![image](https://github.com/TA-Grant1101/g2g/assets/132425270/5d76706f-a39f-4a5a-b0e5-84552f7f6040)
-double check it has been added to the database if you like:
![image](https://github.com/TA-Grant1101/g2g/assets/132425270/5df83450-6cd4-4460-80d3-315a77420e40)

Great! Let's try to add a customer!

1. Navigate to http://localhost:8080/register. we're going to make a POST request to add a new customer. here is a sample customer we can add:
{
    "name": "sad",
    "email": "sad@example.com",
    "role": "user",
    "pwd": "12345",
    "createDt": "2024-04-14",
    "bookLibrary": null,
    "wishlist": null
}
2. paste that customer into the body of the request and press send!
![image](https://github.com/TA-Grant1101/g2g/assets/132425270/15a1cec4-5627-4cab-b942-68713d36b2a3)
NOTE - THIS HAS BROKEN THE ABILITY TO BLOCK DUPLICATE RECORDS - FIX IS COMING IN FRONTEND (once registration is completed)

feel free to test the other routes as well, DELETE book should also work!
